File: Source/Orleankka.Tests/Actors/Scenarios/Request_response.cs
using System;
using System.Linq;

using NUnit.Framework;

namespace Orleankka.Actors.Scenarios
{
    [TestFixture]
    public class Request_response
    {
        static readonly IActorSystem system = ActorSystem.Instance;

        [Test]
        public async void Client_to_actor()
        {
            var actor = system.FreshActorOf<ITestActor>();
            
            await actor.Tell(new SetText("c-a"));
            Assert.AreEqual("c-a", await actor.Ask(new GetText()));
        }

        [Test]
        public async void Actor_to_actor()
        {
            var one = system.FreshActorOf<ITestInsideActor>();
            var another = system.FreshActorOf<ITestActor>();

            await one.Tell(new DoTell(another, new SetText("a-a")));
            Assert.AreEqual("a-a", await one.Ask(new DoAsk(another, new GetText())));
        }
    }
}

File: Source/Orleankka.FSharp.Tests/RequestResponse.fs
module Orleankka.FSharp.RequestResponse

open NUnit.Framework
open System
open Orleankka
open Orleankka.FSharp
open Orleankka.FSharp.Tests.Infrastructure

type Message = 
   | Greet of string
   | Hi

type ITestActor = 
    inherit IActorGrain

type TestActor() = 
   inherit ActorGrain<Message>()
   interface ITestActor

   override this.Receive message = task {
      match message with
      | Greet who -> return response(sprintf "Receive Hello %s" who)
      | Hi        -> return response(sprintf "Receive Hi")
   }

[<TestFixture>]
[<RequiresSilo>]
type Tests() =   
   [<DefaultValue>] val mutable system: IActorSystem

   [<SetUp>]
   member this.SetUp() = 
      this.system <- TestActorSystem.instance
    
   [<Test>]
   member this.``Actor<T> should throws an exception when input message type is different then T type.``() = 
      let actor = ActorSystem.actorOf<ITestActor>(this.system, "test")
      match Task.run(fun _ -> task {return! actor <? "request msg"}) with
      | Choice1Of2 result -> Assert.Fail("actor was able to handle unspecified message type.")
      | Choice2Of2 ex     -> Assert.IsInstanceOf(typeof<Exception>, ex.InnerException)